sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,206 @@
+// must be kept in sync with the data slice.
+// See https://golang.org/pkg/container/heap/#example__priorityQueue
+type PriorityQueue struct {
+	// data is the internal structure that holds the queue, and is operated on by
+	// heap functions
+	data pQueue
+
+	// dataMap represents all the items in the queue, with unique indexes, used
+	// for finding specific items. dataMap is kept in sync with the data slice
+	dataMap map[string]*Item
+
+	// mapMutex is used to facilitate read/write locks on the dataMap
+	dataMutex sync.Mutex
+}
+
+// pQueue is the internal data structure used to satisfy heap.Interface. This
+// prevents users from calling Pop and Push heap methods directly
+type pQueue []*Item
+
+var _ heap.Interface = &pQueue{}
+var _ sort.Interface = &pQueue{}
+
+// Item is something managed in the priority queue
+type Item struct {
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
+
+	// index is an internal value used by the heap package, and should not be
+	// modified by any consumer of the priority queue
+	index int
+}
+
+// Pop pops the highest priority item from the queue. This is a
+// wrapper/convenience method that calls heap.Pop, so consumers do not need to
+// invoke heap functions directly
+func (pq *PriorityQueue) Pop() (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(&pq.data, clone)
+	return nil
+}
+
+// PopByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	item, ok := pq.dataMap[key]
+	if !ok {
+		return nil, newErrItemNotFound(key)
```

Comment:
I generally prefer to `return nil, nil` instead of a not-found error, makes it easier on the caller such that they don't need to check for a specific error type, just a nil *Item. What do you think?

Reply:
That's probably fine, it will do away with `ErrItemNotFound` completely

Reply:
Done in 41ffc88d1